### PR TITLE
Signed install

### PR DESCRIPTION
**Pull request by skv-headless**
based on https://community.developer.atlassian.com/t/action-required-atlassian-connect-installation-lifecycle-security-improvements/49046

- [x] need more testing
- [x] compare with ACE implementation

### CODE REVIEW

File: lib/atlassian-jwt-authentication/verify.rb [hunk 1]
```diff
@@ -49,9 +49,20 @@ def verify
         return false
       end
 
+      if encoding_data['alg'] == 'RS256'
```
Comment (skv-headless) [inline lib/atlassian-jwt-authentication/verify.rb]: I think that condition is not enough. We also need to check descriptor. Not sure if we can do that

ACE reference https://bitbucket.org/atlassian/atlassian-connect-express/src/4bb8249ec63241ba6b22bdfcf28759362c54dcc9/lib/middleware/verify-installation.js#lines-88

Comment (pawelniewie) [reply lib/atlassian-jwt-authentication/verify.rb]: You mean to know if we are using signed installs or not? I think we can add flag to the configuration that would solve that, WDYT?

Comment (skv-headless) [reply lib/atlassian-jwt-authentication/verify.rb]: Yeah. I think adding flag is the only option

**Comment by stefan-hc** [comment]
Looks good to me. You've asked on slack about possible things to break/test, rough list:
* test with two dev Jira instances having app already installed: let one (**J1**) have installed 'previous' version of app without these changes, second one (**J2**) already with these changes, then run the backend for both instances:
* both should be able to uninstall -> install with **old** descriptor -> uninstall
* **J2** should be able to disable -> enable (using the same app version) - _I don't expect this to be affected, but better check if it still works_
* on **J1**: test that app upgrade (installation of the new version) does work
* test unavailability of 'connect-install-keys.atlassian.com' host (will the app+jira provide meaningful error to the customer? will the app logs contain useful message for the app supporters?)
* anything else? :thinking:

**Comment by skv-headless** [comment]
@stefan-hc 

> * **J2** should be able to disable -> enable (using the same app version) - _I don't expect this to be affected, but better check if it still works_

didn't understand 😞  what should I disable -> enable?

> * test unavailability of 'connect-install-keys.atlassian.com' host (will the app+jira provide meaningful error to the customer? will the app logs contain useful message for the app supporters?)

Added log. The error message looks like this

Screenshot 2021-07-29 at 11 50 56

**Comment by stefan-hc** [comment]
> what should I disable -> enable?

Good question. I believe (but my memory might fail me) that at some point in the past there was more buttons in "Manage your apps" page in Jira: not only "uninstall" "install" "subscribe" etc, but also "disable" "enable". If you check Atlassian documentation there are even some mentions about app disabling/enabling: https://support.atlassian.com/jira-cloud-administration/docs/integrate-apps/

Might be it is copy-paste from Jira Server/DataCenter (where you can disable arbitrary apps, heck, you can disable it's modules one by one), might be that Atlassian has removed Enabling/Disabling function from Jira Cloud...

Perhaps "subscribe / unsubscribe" is current name of "enable / disable". What I suggest:
* play a bit with subscribe / unsubscribe - I believe these might call backend on "/enabled" "/disabled" endpoints (instead of /installed /uninstalled). I suppose nothing has changed from the Atlassian side, but it nonetheless might be good idea to check happy path
* if subscribe/unsubscribe is doing something completely different - ignore my "J2 should be able to disable -> enable" request: it has totally no sense.

Ok for you?

**Comment by skv-headless** [comment]
> > what should I disable -> enable?
> 
> Good question. I believe (but my memory might fail me) that at some point in the past there was more buttons in "Manage your apps" page in Jira: not only "uninstall" "install" "subscribe" etc, but also "disable" "enable". If you check Atlassian documentation there are even some mentions about app disabling/enabling: https://support.atlassian.com/jira-cloud-administration/docs/integrate-apps/
> 
> Might be it is copy-paste from Jira Server/DataCenter (where you can disable arbitrary apps, heck, you can disable it's modules one by one), might be that Atlassian has removed Enabling/Disabling function from Jira Cloud...
> 
> Perhaps "subscribe / unsubscribe" is current name of "enable / disable". What I suggest:
> 
> * play a bit with subscribe / unsubscribe - I believe these might call backend on "/enabled" "/disabled" endpoints (instead of /installed /uninstalled). I suppose nothing has changed from the Atlassian side, but it nonetheless might be good idea to check happy path
> * if subscribe/unsubscribe is doing something completely different - ignore my "J2 should be able to disable -> enable" request: it has totally no sense.
> 
> Ok for you?

Tested on staging. Everything works fine for me

**Comment by stefan-hc** [comment]
Awesome.

**Review by pawelniewie — APPROVED** [review]

**Review by bulinutza — APPROVED** [review]
Happy to see more contributors to this gem!

**Pull request merged**